cmd2/cmd2.py

```diff
@@ -526,6 +527,11 @@ def __init__(self, completekey: str='tab', stdin=None, stdout=None, persistent_h
         # This determines if a non-zero exit code should be used when exiting the application
         self.exit_code = None
 
+        # This lock should be acquired before doing any asynchronous changes to the terminal to
+        # ensure the updates to the terminal don't interfere with the input being typed. It can be
+        # acquired any time there is a readline prompt on screen.
+        self.terminal_lock = threading.RLock()
```

Comment:
@kotfu I’d appreciate your review of this PR especially with an eye towards making sure nothing here would negatively impact any of your existing cmd2 apps.